src/services/exportInfoMap.ts

```diff
@@ -77,6 +89,30 @@ namespace ts {
                     cache.clear();
                     usableByFileName = importingFile;
                 }
+
+                let packageName;
+                if (moduleFile) {
+                    const nodeModulesIndex = moduleFile.fileName.indexOf(nodeModulesPathPart);
+                    if (nodeModulesIndex !== -1) {
+                        const isTypes = moduleFile.fileName.substring(nodeModulesIndex + nodeModulesPathPart.length).indexOf("@types/") === 0;
+                        const nextSlash = moduleFile.fileName.indexOf(directorySeparator, nodeModulesIndex + nodeModulesPathPart.length + (isTypes ? "@types/".length : 0) + 1);
+                        packageName = moduleFile.fileName.substring(nodeModulesIndex + nodeModulesPathPart.length + (isTypes ? "@types/".length : 0), nextSlash);

```

Comment:
Namespaced `@types` package names would need further processing, no?

Reply:
No—the `packageName` doesn’t actually need to be unmangled because it’s only for comparing directory names to directory names. (I guess `packageName` is a slight misnomer in this way.) But you’re right that non-@types scopes need to be handled.

Reply:
Weeeell, technically I guess you could end up with typings in `packages/a/node_modules/@types/foo__bar` shadowing typings in `node_modules/@foo/bar`. So yeah. Argh.